### PR TITLE
many: consider asserted components when building a seed

### PR DESCRIPTION
**Pull request by alfonsosanchezbeato**
Consider asserted components when building a seed, so they can be specified as local files but considered asserted if matching assertions are found in the store.

### CODE REVIEW

**Comment by alfonsosanchezbeato** [comment]
A bit drafty yet as some unit tests are missing.

File: asserts/snapasserts/snapasserts.go [hunk 1]
```diff
@@ -379,36 +444,42 @@ func FetchSnapAssertions(f asserts.Fetcher, snapSHA3_384, provenance string) err
...
-// FetchComponentAssertions fetches the assertions matching the information
-// described in the given SideInfo and ComponentSideInfo using the given
-// fetcher.
-func FetchComponentAssertions(f asserts.Fetcher, si *snap.SideInfo, csi *snap.ComponentSideInfo, hash, provenance string) error {
-	// for now starting from the snap-resource-revision will get us all other relevant assertions
+func FetchResourceRevisionAssertion(f asserts.Fetcher, si *snap.SideInfo, compName, hash, provenance string) error {
 	ref := &asserts.Ref{
 		Type:       asserts.SnapResourceRevisionType,
-		PrimaryKey: []string{si.SnapID, csi.Component.ComponentName, hash},
+		PrimaryKey: []string{si.SnapID, compName, hash},
```
Comment (andrewphelpsj) [inline asserts/snapasserts/snapasserts.go:450]: Rename `compName` to `resourceName` to match function signature?

File: asserts/snapasserts/snapasserts.go [hunk 2]
```diff
@@ -379,36 +444,42 @@ func FetchSnapAssertions(f asserts.Fetcher, snapSHA3_384, provenance string) err
...
-		return err
-	}
-
-	// fetch the snap-resource-pair as well
-	ref = &asserts.Ref{
+func FetchResourcePairAssertion(f asserts.Fetcher, si *snap.SideInfo, compName string, compRev snap.Revision, provenance string) error {
+	ref := &asserts.Ref{
 		Type:       asserts.SnapResourcePairType,
-		PrimaryKey: []string{si.SnapID, csi.Component.ComponentName, csi.Revision.String(), si.Revision.String()},
+		PrimaryKey: []string{si.SnapID, compName, compRev.String(), si.Revision.String()},
```
Comment (andrewphelpsj) [inline asserts/snapasserts/snapasserts.go:461]: Same suggestion here

File: seed/seedwriter/writer.go [hunk 3]
```diff
@@ -784,32 +763,23 @@ func (c byCompName) Len() int           { return len(c) }
...
 			return fmt.Errorf("component %s is not defined by snap %s",
-				compInfo.Component.ComponentName, sn.SnapName())
+				seedComp.ComponentName, sn.SnapName())
 		}
 		// and if types match
-		if compInSnap.Type != compInfo.Type {
+		if compInSnap.Type != seedComp.Info.Type {
 			return fmt.Errorf("component %s has type %s while snap %s defines type %s for it",
-				compInfo.Component.ComponentName, compInfo.Type,
+				seedComp.ComponentName, seedComp.Info.Type,
```
Comment (andrewphelpsj) [inline seed/seedwriter/writer.go:777]: I think both of these branches are missing tests.

**Review by andrewphelpsj — COMMENTED** [review]
Thanks, some small suggestions

File: seed/seedwriter/writer.go (hunk 3 above)
Comment (alfonsosanchezbeato) [reply seed/seedwriter/writer.go:777]: There were tests in `image/image_test.go` but it is true that there were no tests in the `seedwriter` package, I've added them now.

**Review by alfonsosanchezbeato — COMMENTED** [review]
@andrewphelpsj thanks for the review, I've addressed your comments now

File: image/image_linux.go [hunk 4]
```diff
@@ -451,18 +451,31 @@ func (s *imageSeeder) validateSnapArchs(snaps []*seedwriter.SeedSnap) error {
...
+		ci, err := readComponentInfoFromCont(path)
+		if err != nil {
+			return nil, err
+		}
+		cinfos[path] = ci
+	}
+
 	snaps := make(localSnapRefs)
 	for _, sn := range localSnaps {
 		// TODO try to get assertions for local components
```
Comment (pedronis) [inline image/image_linux.go:471]: should this TODO go away now?

File: asserts/snapasserts/snapasserts.go [hunk 5]
```diff
@@ -365,6 +387,49 @@ func SideInfoFromSnapAssertions(snapDecl *asserts.SnapDeclaration, snapRev *asse
...
+		}
+		a = cands[0]
+	}
+
+	resRev := a.(*asserts.SnapResourceRevision)
+
+	if resRev.ResourceSize() != resSize {
+		return nil, fmt.Errorf("resource %q does not have the expected size according to signatures (broken or tampered): %d != %d", resName, resSize, resRev.ResourceSize())
+	}
+
```
Comment (pedronis) [inline asserts/snapasserts/snapasserts.go:422]: if you compare with the same function for snaps, we need to run crossCheckResourceProvenance  here as well at this point

**Review by pedronis — CHANGES_REQUESTED** [review]
did a first pass, one main comment about checking that the resource revision is actually within the revision authority constraints of the snap if applicable

File: image/image_linux.go (hunk 4 above)
Comment (alfonsosanchezbeato) [reply image/image_linux.go]: Removed

**Comment by alfonsosanchezbeato** [comment]
@pedronis @andrewphelpsj thanks for the reviews, this is ready for another pass now

**Review by pedronis — APPROVED** [review]
thank you

**Comment by alfonsosanchezbeato** [comment]
I've squashed the commits now

**Pull request merged**